state/src/replicationv2/snapshot.rs

```diff
@@ -157,53 +171,118 @@ impl StateMachine {
         fs::remove_file(path).map_err(err_str!(ReplicationV2Error::Snapshot))?;
         Ok(())
     }
+
+    // ----------------
+    // | Installation |
+    // ----------------
+
+    /// Unzip a snapshot and open a DB from the data
+    pub(crate) async fn open_snap_db(&self) -> Result<DB, ReplicationV2Error> {
+        // Open the file
+        let zip_path = self.snapshot_archive_path();
+        let file = tokio::fs::File::open(zip_path)
+            .await
+            .map_err(err_str!(ReplicationV2Error::Snapshot))?;
+        self.open_snap_db_with_file(file).await
+    }
+
+    /// Open a snapshot DB from a given file
+    pub(crate) async fn open_snap_db_with_file(
+        &self,
+        file: tokio::fs::File,
+    ) -> Result<DB, ReplicationV2Error> {
+        // Unzip the snapshot
+        let dest_path = self.snapshot_data_path();
+        let db_path = dest_path.clone();
+        let std_file = file
+            .try_into_std()
+            .map_err(|_| ReplicationV2Error::Snapshot(ERR_CONVERT_FILE.to_string()))?;
+        let job = tokio::task::spawn_blocking(move || Self::unzip_data_file(&std_file, &dest_path));
+        job.await.map_err(|_| ReplicationV2Error::Snapshot(ERR_AWAIT_BUILD.to_string()))??;
+
+        // Open the DB
+        let path = db_path.to_str().unwrap().to_string();
+        let db_config = DbConfig::new_with_path(&path);
+        let db = DB::new(&db_config).map_err(ReplicationV2Error::Storage)?;
+
+        Ok(db)
+    }
+
+    /// Update the local DB using a snapshot DB
+    pub async fn update_from_snapshot(
+        &mut self,
+        meta: &SnapshotMeta<NodeId, Node>,
+        snapshot_db: DB,
+    ) -> Result<(), ReplicationV2Error> {
+        self.last_applied_log = meta.last_log_id;
+        self.last_membership = meta.last_membership.clone();
```

Comment:
do we want to carry over membership from the snapshot? I could see this changing drastically between when a snapshot was taken and when it is applied

Reply:
ah, nvm - this is just the `last_membership` metadata right, not the nodes actual view of the membership (which is in an excluded table)?

Reply:
yeah the last membership change the node is aware of. we do want this; a leader will only send a snapshot if a node is far enough behind to not replicate logs directly. So yes, it may change a lot but only to a more recent value.